### ISSUE TITLE
iw3 dev branch: Frame/Depth Mismatch in `bind_batch_frame_callback` Causes Frame Jumps

### ISSUE THREAD

**Issue by taowen** [nagadomi/nunif#420]
## Summary

When processing video frames in batch mode with `bind_batch_frame_callback` in `iw3/utils.py`, there is a mismatch between the number of frames stored in `src_queue` and the number of depth maps returned by `depth_model.minmax_normalize`. This leads to incorrect pairing of frames and depth maps, causing visible frame jumps and discontinuities in the output video.

---

## Steps to Reproduce

1. Run the video processing pipeline with batch mode enabled and EMA normalization active.
2. Observe the logs and output video. Occasionally, the output frames will "jump" or appear out of order.
3. The logs will show repeated warnings like:

   ```
   [BUG][bind_batch_frame_callback][_postprocess] src_queue and depth_list length mismatch! src_queue=8, depth_list=0
   [BUG][bind_batch_frame_callback][_postprocess] src_queue pts=[...]
   [BUG][bind_batch_frame_callback][_postprocess] depth_list count=0
   ```

---

### Code Location

**File:** `iw3/utils.py`  
**Function:** `bind_batch_frame_callback`

#### Relevant Code

```python
src_queue = []

def _postprocess(depth_list, device_index):
    if len(depth_list) != len(src_queue):
        print(f"[BUG][bind_batch_frame_callback][_postprocess] src_queue and depth_list length mismatch! src_queue={len(src_queue)}, depth_list={len(depth_list)}")
        print(f"[BUG][bind_batch_frame_callback][_postprocess] src_queue pts={[x[1] for x in src_queue]}")
        print(f"[BUG][bind_batch_frame_callback][_postprocess] depth_list count={len(depth_list)}")
    # ... existing code ...

@torch.inference_mode()
def _batch_frame_callback(x, pts, flush):
    if flush:
        depth_list = depth_model.flush_minmax_normalize()
        if len(depth_list) != len(src_queue):
            print(f"[BUG][bind_batch_frame_callback][flush] src_queue and depth_list length mismatch! src_queue={len(src_queue)}, depth_list={len(depth_list)}")
        return _postprocess(depth_list, device_index=0)

    device_index = args.state["devices"].index(x.device)
    with preprocess_lock[device_index]:
        x = preprocess_image(x, args)

    if lookahead_enabled:
        x_cpu = torch.clamp(x.permute(0, 2, 3, 1) * 255.0, 0, 255).to(torch.uint8).cpu()
        for x_, pts_ in zip(x_cpu, pts):
            src_queue.append((x_, pts_))
    else:
        src_queue.append((x, pts))

    with depth_lock[device_index]:
        depths = depth_model.infer(x, tta=args.tta, low_vram=args.low_vram,
                                   enable_amp=not args.disable_amp,
                                   edge_dilation=args.edge_dilation,
                                   depth_aa=args.depth_aa)
        reset_ema = [t in segment_pts for t in pts]
        depth_list = depth_model.minmax_normalize(depths, reset_ema=reset_ema)
        return _postprocess(depth_list, device_index)
```

### Explanation

- **src_queue** accumulates every input frame (and its pts) as soon as a batch is processed.
- **depth_model.minmax_normalize** only returns a depth map when its internal buffer is filled (due to EMA logic). Otherwise, it returns `None` or an empty list.
- This means that for many calls, `src_queue` grows, but `depth_list` is empty.
- When a flush occurs, or when the buffer is finally filled, a batch of depth maps is returned, but the number of depth maps may not match the number of frames in `src_queue`.
- This leads to a mismatch: frames and depth maps are not paired correctly, causing output frames to jump or appear out of order.

---

## Impact

- **User-visible:** Output video has frames that jump or are out of sequence.
- **Technical:** Internal state (`src_queue`) and output (`depth_list`) are not synchronized, leading to data corruption.

---

I asked cursor to write the bug report for me. What I noticed is “Scene Boundary Detection" in the dev branch is causing the video frame discontinuity.  Here is the setup I am using

![Image](https://github.com/user-attachments/assets/f4d71001-00bf-4cdc-a192-aa8d9889ea6e)

**Comment by nagadomi** [comment]
The number of frames added to `src_queue` and the number of frames returned by `depth_model.minmax_normalize()`+`depth_model.flush_minmax_normalize()` should eventually match. Otherwise, some runtime error (IndexError or None type error) occur.
I tested it with the following changes, but the result was always 0.
```python
     def _batch_frame_callback(x, pts, flush):
         if flush:
             depth_list = depth_model.flush_minmax_normalize()
-            return _postprocess(depth_list, device_index=0)
+            results = _postprocess(depth_list, device_index=0)
+            print("flush: len(src_queue)=", len(src_queue))  # len(src_queue) must be zero at the end of video processing.
+            return results
```

----

If source-depth pairing fails, it is probably due to the worker threads changing the `pop` order from `src_queue`. I was aware of this risk, but left it alone since I had never seen it occur.
If this is the cause, the issue should not occur when `Worker Threads` is set to `0`.

Depth estimation can be parallelized. But the other parts need to be executed sequentially, I think.

**Comment by nagadomi** [comment]
I plan to make postprocess (minmax_normalize and stereo generation) sequential.
Although the callback itself is called sequentially, but I suspect flush_minmax_normalize causes timing differences across threads, which changes the order of postprocess.

**Comment by nagadomi** [comment]
I have not reproduced or confirmed the fix for the issue, but I believe it was fixed in `fix_iw3_batch` branch (PR #421).
If this change has fixed the reported issue, I will merge it. FPS might be slightly lower.

**Comment by taowen** [comment]
I will check it tonight, thank you for your quick fix

**Comment by taowen** [comment]
I confirm the pull request fixed the bug. You can check the converted video here: https://github.com/taowen/iw3-bug-report/tree/main/Frame%20Depth%20Mismatch%20in%20bind_batch_frame_callback%20Causes%20Frame%20Jumps

larger batch size will make the bug more obvious

**Comment by nagadomi** [comment]
Thank you for checking. The fix has been merged into the dev branch.

**Comment by nagadomi** [comment]
As for performance, I will try to improve it later if possible.

**Comment by nagadomi** [comment]
I reworked the changes from #421 in #424.
I tested it on my side, but since the previous change was reverted, let me know if the frame order is broken again.

**Comment by taowen** [comment]
I used #424 and converted the same video https://github.com/taowen/iw3-bug-report/tree/main/Frame%20Depth%20Mismatch%20in%20bind_batch_frame_callback%20Causes%20Frame%20Jumps

It is working fine, good job

**Comment by nagadomi** [comment]
Thank you for checking.
